presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java

```diff
@@ -355,55 +356,44 @@ public Builder addEquality(Expression expression)
 
         public Builder addEquality(Expression expression1, Expression expression2)
         {
-            checkArgument(!expression1.equals(expression2), "need to provide equality between different expressions");
             checkArgument(DeterminismEvaluator.isDeterministic(expression1), "Expression must be deterministic: " + expression1);
             checkArgument(DeterminismEvaluator.isDeterministic(expression2), "Expression must be deterministic: " + expression2);
 
-            Expression canonical1 = canonicalize(expression1);
-            Expression canonical2 = canonicalize(expression2);
-
-            if (!canonical1.equals(canonical2)) {
-                map.put(canonical1, canonical2);
-                reverseMap.put(canonical2, canonical1);
-            }
+            unionFindSet.findAndUnion(expression1, expression2);
```

Comment:
that's just a library i'm using. if the name is a problem, we should fix it in another PR.